### PR TITLE
count_consonants

### LINKED ISSUE #8: count_consonants
# Count Consonants in a String

Write a function that takes in a string and returns the number of consonants (all alphabetic characters except vowels) in the string.

### PR DESCRIPTION
**Pull request by Norbert250**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_count_consonants.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on 03 01 2025
+
+@author: Norbert Ndayisenga
+"""
```
Comment (mohamad-755) [inline solutions/tests/test_count_consonants.py]: I think this current docstring doesn't provide enough information about the test and function.
You can write a brief and concise explanation about the function and the test.

File: solutions/tests/test_count_consonants.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        self.assertEqual(count_consonants("aeiouAEIOU"), 0)
+
+    def test_all_consonants(self):
+        """It should evaluate a string of only consonants correctly."""
+        self.assertEqual(count_consonants("bcdfghjklmnpqrstvwxyz"), 21)
+
+    def test_mixed_characters(self):
+        """It should handle mixed cases of vowels, consonants, and symbols."""
+        self.assertEqual(count_consonants("Hello, World!"), 7)
+
```
Comment (mohamad-755) [inline solutions/tests/test_count_consonants.py:31]: You should add tests to verify that the function raises an exception or handles invalid input appropriately (for example, when the input is not a string).

File: solutions/tests/test_count_consonants.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    def test_all_vowels(self):
+        """It should evaluate a string of only vowels to 0."""
+        self.assertEqual(count_consonants("aeiouAEIOU"), 0)
+
+    def test_all_consonants(self):
+        """It should evaluate a string of only consonants correctly."""
+        self.assertEqual(count_consonants("bcdfghjklmnpqrstvwxyz"), 21)
+
+    def test_mixed_characters(self):
+        """It should handle mixed cases of vowels, consonants, and symbols."""
```
Comment (mohamad-755) [inline solutions/tests/test_count_consonants.py]: I think the name of this function should be "test_mixed_case_and_special_characters" to specify exactly what's being tested.

File: solutions/count_consonants.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on 03 01 2024
+
+@author: Norbert Ndayisenga
```
Comment (mohamad-755) [inline solutions/count_consonants.py]: same as the test.

**Review by mohamad-755 — CHANGES_REQUESTED** [review]
There are some changes should be made.

**Review by mohamad-755 — APPROVED** [review]
The function is well-designed, efficient, and easy to understand.
